File: QAP/Java/Individual.java
package com.qap;

import java.util.Random;

public class Individual {
    /**
     * Vector de cromosomas cuyos elementos no se repiten
     */
    private int[] chromosomes;

    /**
     * Fitness del individuo
     */
    private int fitness;

    /**
     * Constructor
     *
     * @param size tamaño del vector de cromosomas del individuo
     * @param empty true si el vector de cromosomas se inicializa a -1, false si se genera una
     *              solución aleatoria
     * @return individuo
     */
    public Individual(int size, boolean empty){
        this.fitness = Integer.MAX_VALUE;
        this.chromosomes = new int[size];

        if(!empty){
            for(int i=0; i<size; ++i){
                chromosomes[i] = i;
            }

            shuffleChromosomes();
        }

        else{
            for(int i=0; i<size; ++i){
                chromosomes[i] = -1;
            }
        }
    }

    /**
     * Constructor de copia
     *
     * @param i individuo a copiar
     * @return copia del individuo
     */
    public Individual(Individual i){
        this.fitness = i.fitness;
        this.chromosomes = new int[i.chromosomes.length];
        System.arraycopy( i.chromosomes, 0, this.chromosomes, 0, i.chromosomes.length );
    }

    /**
     * Genera una solución aleatoria del vector de cromosomas del individuo
     *
     */
    private void shuffleChromosomes(){
        Random rnd = new Random();
        for(int i=this.chromosomes.length-1; i>0; --i){
            int index = rnd.nextInt(i+1);
            int aux = this.chromosomes[index];
            this.chromosomes[index] = this.chromosomes[i];
            this.chromosomes[i] = aux;
        }
    }

    /**
     * Calcula el fitness del individuo
     *
     */
    public void calculateFitness(){
        int totalFitness = 0;
        for(int i=0; i<Reader.getFlowMatrix().length; ++i){
            for(int j=0; j<Reader.getFlowMatrix().length; ++j){
                int chromosomeI = this.chromosomes[i];
                int chromosomeJ = this.chromosomes[j];

                totalFitness += Reader.getFlowMatrix()[i][j] * Reader.getDistanceMatrix()[chromosomeI][chromosomeJ];
            }
        }

        if(totalFitness < 44759294){
            System.out.println("Error al calcular el fitness");
        }

        this.fitness = totalFitness;
    }

    /**
     * Comprueba si un individuo es igual a otro
     *
     * @param c cromosomas del individuo a evaluar
     * @return true si los cromosomas del individuo son iguales, false en caso contrario
     */
    public void calculateFitnessBaldwinian(){
        Individual best = new Individual(this);
        int bestFitness = best.getFitness();

        for(int i=0; i<best.getChromosomes().length; ++i) {
            for (int j = i + 1; j < best.getChromosomes().length; ++j) {
                Individual T = new Individual(best);
                T.setChromosome(best.getChromosomes()[j], i);
                T.setChromosome(best.getChromosomes()[i], j);
                int optFitness = T.calculateOptFitness(T.getChromosomes());

                if (optFitness < best.getFitness()) {
                    best.fitness = optFitness;
                    best.chromosomes = T.getChromosomes();
                    this.fitness = optFitness;
                }
            }
        }
    }

    /**
     * Búsqueda de óptimo local para la versión Lamarckiana
     *
     */
    public void calculateFitnessLamarckian(){
        Individual best = new Individual(this);

        for(int i=0; i<best.getChromosomes().length; ++i){
            for(int j=i+1; j<best.getChromosomes().length; ++j){
                Individual T = new Individual(best);
                T.setChromosome(best.getChromosomes()[j],i);
                T.setChromosome(best.getChromosomes()[i],j);
                int optFitness = T.calculateOptFitness(T.getChromosomes());

                if(optFitness < best.getFitness()){
                    best.fitness = optFitness;
                    best.chromosomes = T.getChromosomes();
                    this.fitness = optFitness;
                    this.chromosomes = best.getChromosomes();
                }
            }
        }
    }

    /**
     * Calcula el fitness de un vector de cromosomas
     *
     * @param c vector de cromosomas
     * @return fitness calculado del vector
     */
    private int calculateOptFitness(int[] c){
        int totalFitness = 0;
        for(int i=0; i<Reader.getFlowMatrix().length; ++i){
            for(int j=0; j<Reader.getFlowMatrix().length; ++j){
                int chromosomeI = c[i];
                int chromosomeJ = c[j];

                totalFitness += Reader.getFlowMatrix()[i][j] * Reader.getDistanceMatrix()[chromosomeI][chromosomeJ];
            }
        }

        if(totalFitness < 44759294){
            System.out.println("Error al calcular el fitness");
        }

        return totalFitness;
    }

    /**
     * Comprueba si un individuo es igual a otro
     *
     * @param c cromosomas del individuo a evaluar
     * @return true si los cromosomas del individuo son iguales, false en caso contrario
     */
    public boolean equals(int[] c){
        for(int i=0; i<this.chromosomes.length; ++i){
            if(this.chromosomes[i] != c[i])
                return false;
        }
        return true;
    }

    /**
     * Intercambia dos cromosomas
     *
     * @param index1 índice del cromosoma
     * @param index2 índice de cromosoma
     */
    public void swap(int index1, int index2){
        int aux = this.chromosomes[index1];
        this.chromosomes[index1] = this.chromosomes[index2];
        this.chromosomes[index2] = aux;
    }

    /**
     * Devuelve los cromosomas del individuo
     *
     * @return cromosomas del individuo
     */
    public int[] getChromosomes() {
        return chromosomes;
    }

    /**
     * Sustituye un cromosoma en el individuo
     *
     * @param value valor del cromosoma
     * @param index índice en el que se va a insertar
     */
    public void setChromosome(int value, int index){
        this.chromosomes[index] = value;
    }

    /**
     * Devuelve el fitness de un individuo
     *
     * @return fitness del individuo
     */
    public int getFitness() {
        return fitness;
    }

    /**
     * Comprueba si un individuo contiene un cromosoma
     *
     * @param chromosome cromosoma
     * @return true si el individuo contiene el cromosoma, false en caso contrario
     */
    public boolean containsChromosome(int chromosome){
        for(int c : this.chromosomes){
            if(c == chromosome){
                return true;
            }
        }

        return false;
    }
}

File: QAP/Python/individual.py
import math
import random

class Individual:

	def __init__(self, size, individual=None):
		self.size = size
		self.fitness = math.inf
		if individual:
			self.copy_individual(individual)
		else:
			self.generate_random_solution()

	def generate_random_solution(self):
		""" Genera una solución aleatoria en el vector de cromosomas del individuo """

		self.chromosomes = list(range(self.size))
		random.shuffle(self.chromosomes)

	def copy_individual(self, individual):
		""" Constructor de copia de un individuo """

		self.chromosomes = list(range(self.size))
		for i in range(self.size):
			self.chromosomes[i] = individual[i]

	def calculate_fitness(self, flow_matrix, distance_matrix):
		""" Calcula el fitness de un individuo """

		total_fitness = 0
		matrix_size = range(len(flow_matrix))
		for i in matrix_size:
			for j in matrix_size:
				chromosomes_i = self.chromosomes[i]
				chromosomes_j = self.chromosomes[j]
				total_fitness += flow_matrix[i][j] * distance_matrix[chromosomes_i][chromosomes_j]

		# Asegura que el fitness no supera el record mundial actual
		assert(total_fitness >= 44759294)

		self.fitness = total_fitness

	def calculate_fitness_baldwiniana(self, flow_matrix, distance_matrix):
		""" Calcula el fitness de un individuo en la versión baldwiniana """

		best = Individual(self.size, self.chromosomes)
		
		for i in range(self.size):
			for j in range(i+1, self.size):
				T = Individual(best.size, best.chromosomes)
				T.chromosomes[i] = best.chromosomes[j]
				T.chromosomes[j] = best.chromosomes[i]
				T.calculate_fitness(flow_matrix, distance_matrix)

				if T.fitness < best.fitness:
					best = T
					self.fitness = best.fitness

	def calculate_fitness_lamarckiana(self, flow_matrix, distance_matrix):
		""" Calcula el fitness de un individuo en la versión lamarckiana"""

		best = Individual(self.size, self.chromosomes)
		
		for i in range(self.size):
			for j in range(i+1, self.size):
				T = Individual(best.size, best.chromosomes)
				T.chromosomes[i] = best.chromosomes[j]
				T.chromosomes[j] = best.chromosomes[i]
				T.calculate_fitness(flow_matrix, distance_matrix)

				if T.fitness < best.fitness:
					best = T
					self.fitness = best.fitness
					this.chromosomes = best.chromosomes
					
	def swap(self, index1, index2):
		""" Intercambia dos cromosomas de un individuo """

		aux = self.chromosomes[index1]
		self.chromosomes[index1] = self.chromosomes[index2]
		self.chromosomes[index2] = aux